File: VTI/Services/PDFCompilerService.swift
import Foundation

struct PDFCompilationResult: Sendable {
    let pdfURL: URL?
    let log: String
}

struct PDFCompilerService: Sendable {
    func compile(texURL: URL) throws -> PDFCompilationResult {
        let executable = try pdflatexExecutable()
        let outputDirectory = texURL.deletingLastPathComponent()
        let texFileName = texURL.lastPathComponent
        let firstRun = try runPdflatex(
            executable: executable,
            texFileName: texFileName,
            outputDirectory: outputDirectory
        )
        let secondRun = try runPdflatex(
            executable: executable,
            texFileName: texFileName,
            outputDirectory: outputDirectory
        )

        let log = firstRun + "\n\n----- SECOND PASS -----\n\n" + secondRun
        let pdfURL = outputDirectory.appendingPathComponent(texURL.deletingPathExtension().lastPathComponent).appendingPathExtension("pdf")

        guard FileManager.default.fileExists(atPath: pdfURL.path) else {
            throw AppError.pdfCompilationFailed(log)
        }

        return PDFCompilationResult(pdfURL: pdfURL, log: log)
    }

    func pdflatexExecutable() throws -> URL {
        let candidates = [
            "/Library/TeX/texbin/pdflatex",
            "/usr/texbin/pdflatex",
            "/opt/homebrew/bin/pdflatex",
            "/usr/local/bin/pdflatex"
        ]

        if let match = candidates.first(where: { FileManager.default.isExecutableFile(atPath: $0) }) {
            return URL(fileURLWithPath: match)
        }

        throw AppError.pdflatexNotInstalled
    }

    private func runPdflatex(
        executable: URL,
        texFileName: String,
        outputDirectory: URL
    ) throws -> String {
        let process = Process()
        process.executableURL = executable
        process.currentDirectoryURL = outputDirectory
        process.arguments = [
            "-interaction=nonstopmode",
            "-halt-on-error",
            "-file-line-error",
            "-output-directory",
            outputDirectory.path,
            texFileName
        ]

        let outputPipe = Pipe()
        process.standardOutput = outputPipe
        process.standardError = outputPipe

        try process.run()
        process.waitUntilExit()

        let data = outputPipe.fileHandleForReading.readDataToEndOfFile()
        let log = String(data: data, encoding: .utf8) ?? "Nessun log disponibile."
        guard process.terminationStatus == 0 else {
            throw AppError.pdfCompilationFailed(log)
        }
        return log
    }
}

File: VTI.Windows/src/VTI.Core/PDFCompilerService.cs
using System.Diagnostics;
using System.Runtime.InteropServices;
using System.Text;

namespace VTI.Core;

public sealed record PdfCompilationResult(string? PdfPath, string Log);

public sealed class PDFCompilerService
{
    public PdfCompilationResult Compile(string texPath)
    {
        var executable = FindPdflatex();
        var outputDirectory = Path.GetDirectoryName(texPath)!;
        var texFileName = Path.GetFileName(texPath);
        var first = RunPdflatex(executable, texFileName, outputDirectory);
        var second = RunPdflatex(executable, texFileName, outputDirectory);
        var log = first + "\n\n----- SECOND PASS -----\n\n" + second;
        var pdfPath = Path.Combine(outputDirectory, Path.GetFileNameWithoutExtension(texPath) + ".pdf");
        if (!File.Exists(pdfPath))
            throw AppException.PdfCompilationFailed(log);
        return new PdfCompilationResult(pdfPath, log);
    }

    public string FindPdflatex()
    {
        foreach (var candidate in CandidatePaths())
        {
            if (File.Exists(candidate))
                return candidate;
        }
        throw AppException.PdflatexNotInstalled();
    }

    private static IEnumerable<string> CandidatePaths()
    {
        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
        {
            yield return Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles), "MiKTeX", "miktex", "bin", "x64", "pdflatex.exe");
            yield return Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ProgramFilesX86), "MiKTeX", "miktex", "bin", "pdflatex.exe");
            foreach (var year in new[] { "2025", "2024", "2023" })
            {
                yield return Path.Combine("C:\\texlive", year, "bin", "win32", "pdflatex.exe");
                yield return Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), "texlive", year, "bin", "win32", "pdflatex.exe");
            }
        }
        else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
        {
            yield return "/Library/TeX/texbin/pdflatex";
            yield return "/usr/texbin/pdflatex";
            yield return "/opt/homebrew/bin/pdflatex";
            yield return "/usr/local/bin/pdflatex";
        }
        else
        {
            yield return "/usr/bin/pdflatex";
        }

        var pathEnv = Environment.GetEnvironmentVariable("PATH");
        if (string.IsNullOrEmpty(pathEnv)) yield break;
        var sep = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? ';' : ':';
        var name = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "pdflatex.exe" : "pdflatex";
        foreach (var dir in pathEnv.Split(sep, StringSplitOptions.RemoveEmptyEntries))
        {
            var full = Path.Combine(dir.Trim(), name);
            if (File.Exists(full))
                yield return full;
        }
    }

    private static string RunPdflatex(string executable, string texFileName, string outputDirectory)
    {
        var logPath = Path.Combine(outputDirectory, Path.GetFileNameWithoutExtension(texFileName) + ".log");
        var psi = new ProcessStartInfo
        {
            FileName = executable,
            WorkingDirectory = outputDirectory,
            UseShellExecute = false,
            RedirectStandardOutput = false,
            RedirectStandardError = false,
            CreateNoWindow = true,
            ArgumentList =
            {
                "-interaction=nonstopmode",
                "-halt-on-error",
                "-file-line-error",
                "-output-directory",
                outputDirectory,
                texFileName
            }
        };
        using var p = Process.Start(psi) ?? throw AppException.Io("Impossibile avviare pdflatex.");
        p.WaitForExit();
        var log = ReadPdflatexLogFile(logPath);
        if (p.ExitCode != 0)
            throw AppException.PdfCompilationFailed(log);
        return log;
    }

    /// <summary>Reads the transcript written by pdfTeX to <c>basename.log</c> (UTF-8 with tolerant fallback).</summary>
    private static string ReadPdflatexLogFile(string logPath)
    {
        try
        {
            if (!File.Exists(logPath))
                return "(file .log non trovato: " + logPath + ")";
            var bytes = File.ReadAllBytes(logPath);
            return DecodeLogBytes(bytes);
        }
        catch (Exception ex)
        {
            return "Lettura log fallita: " + ex.Message;
        }
    }

    private static string DecodeLogBytes(byte[] bytes)
    {
        if (bytes.Length == 0) return "";
        try
        {
            return new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true).GetString(bytes);
        }
        catch (DecoderFallbackException)
        {
            return Encoding.Latin1.GetString(bytes);
        }
    }
}
